### PR TITLE
Pp 3257 CSRF tokens

### PR DESCRIPTION
**Pull request by rauligs**
## WHAT
- Add CSRF middleware
- Add CSRF token value in forms for set up and confirmation routes

### CODE REVIEW

File: common/middleware/csrf.tests.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
+const path = require('path')
+const sinon = require('sinon')
+const assert = require('assert')
+const proxyquire = require('proxyquire')
+
+describe('CSRF', function () {
+  it('should create a CSRF token', function () {
+    const verify = sinon.stub()
```
Comment (DanailMinchev) [inline common/middleware/csrf.tests.js]: `sinon` is not installed, so we need to install it

File: common/middleware/csrf.tests.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      .withArgs("it's a secret", 'submitted token')
+      .returns(true)
+
+    const create = sinon.stub()
+      .withArgs("it's a secret")
+      .returns('newly-created token')
+
+    const csrf = proxyquire(path.join(__dirname, '/../../common/middleware/csrf.js'),
```
Comment (DanailMinchev) [inline common/middleware/csrf.tests.js]: `proxyquire` is also not installed

File: app/setup/post.controller.tests.js [hunk 3]
```diff
@@ -24,15 +25,44 @@ describe('setup post controller', () => {
     nock.cleanAll()
   })
 
+  describe('when CSRF is not valid', () => {
+    let response
+    let paymentRequestExternalId = 'sdfihsdufh2ff'
```
Comment (DanailMinchev) [inline app/setup/post.controller.tests.js]: probably should be `const`

File: app/setup/post.controller.tests.js [hunk 4]
```diff
@@ -24,15 +25,44 @@ describe('setup post controller', () => {
...
+      expect(response.statusCode).to.equal(400)
+    })
+  })
   describe('when submitting the form for a valid payment request', () => {
     const paymentRequestExternalId = 'sdfihsdufh2e'
     const paymentRequest = paymentFixtures.validPaymentRequest({
       external_id: paymentRequestExternalId
     })
     const formValues = paymentFixtures.validPayer()
+    let csrfSecret = '123'
```
Comment (DanailMinchev) [inline app/setup/post.controller.tests.js]: probably should be `const`

File: app/setup/post.controller.tests.js (hunk 3 above)
Comment (rauligs) [reply app/setup/post.controller.tests.js]: 👍

File: app/setup/post.controller.tests.js (hunk 4 above)
Comment (rauligs) [reply app/setup/post.controller.tests.js]: 👍

File: app/setup/index.js [hunk 5]
```diff
@@ -15,8 +16,8 @@ const paths = {
 }
 
 // Routing
-router.get(paths.index, getController)
-router.post(paths.index, postController)
+router.get(paths.index, ensureSessionHasCsrfSecret, validateAndRefreshCsrf, getController)
+router.post(paths.index, validateAndRefreshCsrf, postController)
```
Comment (DanailMinchev) [inline app/setup/index.js]: Nice optimisation! 👍

**Pull request merged**